### DISCUSSION TITLE
Issue re-authenticating post enabling of 'Security Defaults'

### DISCUSSION THREAD

**Discussion by bossanova808** [abraunegg/onedrive#2235] [Q&A] [answered]
MS applied 'Security Defaults' to our org, meaning we must have 2FA.

So therefore I need to re-auth my OneDrive docker...which I seem unable to do.  As in - the re-auth works fine on the face of it, however it does not appearing to be saving (and no new `refresh.token` is made in the config folder).

```
docker run -it --name onedrive \
>     -v "/home/me/docker/config/onedrive":/onedrive/conf \
>     -v "/media/store03/OneDrive":/onedrive/data \
>     -e "ONEDRIVE_UID=1000" -e "ONEDRIVE_GID=1000" \
>     driveone/onedrive:debian --reauth
Base Args: --monitor
# Launching onedrive
Deleting the saved authentication status ... re-authentication requested
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=blah&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri: https://login.microsoftonline.com/common/oauth2/nativeclient?code=massive_long_code

**Application has been successfully authorised, however no additional command switches were provided.**

Please use 'onedrive --help' for further assistance in regards to running this application.

```

If I then try and actually start the OneDrive container it just gives:

```
usermod: no changes
Base Args: --monitor
# Launching onedrive
Deleting the saved authentication status ... re-authentication requested
Configuring Global Azure AD Endpoints
Authorize this app visiting:

https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=blah&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient

Enter the response uri:
```

...so I am stuck in this loop requiring re-auth.

I think I must have done something wrong, but can't work it out - what am I missing?

Thanks for your work, and any help!

**Comment by abraunegg** [comment, marked as answer]
@bossanova808 
> If I then try and actually start the OneDrive container it just gives:
> ...
> `Deleting the saved authentication status ... re-authentication requested`
> ...

So it appears you are *trying* to restart the re-authed Docker container with `--reauth` ... which, means it will ... re-auth ... and delete the new `refresh.token` ....


To resolve, use 1 process to reauth the Docker container then, restart the Docker container without the `--reauth` command so that the Docker container uses the new|updated `refresh.token`

**Comment by bossanova808** [comment]
@abraunegg   I am only calling the re-auth on the command line invocation (as above) to do the re-auth, once.

My docker-compose, which I then use to start the container (`docker start onedrive`) - does not have re-auth

```
    onedrive:
        container_name: onedrive
        image: driveone/onedrive:debian
        restart: unless-stopped
        environment:
            - ONEDRIVE_UID=${UID}
            - ONEDRIVE_GID=${GUID}
            - ONEDRIVE_DOWNLOADONLY=1
        volumes:
            - ${ONEDRIVE_CONFIG}:/onedrive/conf:rw
            - ${ONEDRIVE_FOLDER}:/onedrive/data:rw
```

**Reply by abraunegg** [reply]
@bossanova808 
However your application output where you say you are just running the Docker container post the reauth is saying something different.

Please double check:
* your Docker commands
* your configuration 'config' file ... have you updated that to include a reauth ask ?

**Comment by bossanova808** [comment]
Yes, I saw that, but I think that is because no token is being saved in the initial step.

My config has no re-auth in it (did not realise that was an option to be honest):

```
# Configuration for OneDrive Linux Client
# This file contains the list of supported configuration fields
# with their default values.
# All values need to be enclosed in quotes
# When changing a config option below, remove the '#' from the start of the line
# For explanations of all config options below see docs/USAGE.md or the man page.
#
# sync_dir = "~/OneDrive"
# skip_file = "~*|.~*|*.tmp"
monitor_interval = "1800"
# skip_dir = ""
# log_dir = "/var/log/onedrive/"
# drive_id = ""
# upload_only = "false"
# check_nomount = "false"
# check_nosync = "false"
download_only = "true"
# disable_notifications = "false"
# disable_upload_validation = "false"
# enable_logging = "false"
# force_http_11 = "false"
# force_http_2 = "false"
# local_first = "false"
# no_remote_delete = "false"
# skip_symlinks = "false"
# debug_https = "false"
# skip_dotfiles = "false"
# dry_run = "false"
# min_notify_changes = "5"
# monitor_log_frequency = "5"
# monitor_fullscan_frequency = "10"
# sync_root_files = "false"
# classify_as_big_delete = "1000"
# user_agent = ""
# remove_source_files = "false"
# skip_dir_strict_match = "false"
# application_id = ""
# resync = "false"
# bypass_data_preservation = "false"
# azure_ad_endpoint = ""
# azure_tenant_id = "common"
# sync_business_shared_folders = "false"
```

And my docker command is `docker-compose up onedrive`  which, after the apparently successful auth, gives:

```
> docker-compose up onedrive
Creating onedrive ... done
Attaching to onedrive
onedrive    | Base Args: --monitor
onedrive    | # We are synchronizing in download-only mode
onedrive    | # Adding --download-only
onedrive    | # Launching onedrive
onedrive    | Configuration file successfully loaded
onedrive    | Configuring Global Azure AD Endpoints
onedrive    | Authorize this app visiting:
onedrive    | 
onedrive    | https://login.microsoftonline.com/common/oauth2/v2.0/authorize?client_id=d50ca740-c83f-4d1b-b616-12c519384f0c&scope=Files.ReadWrite%20Files.ReadWrite.All%20Sites.ReadWrite.All%20offline_access&response_type=code&prompt=login&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient
onedrive    | 
onedrive    | Enter the response uri: Could not initialize the OneDrive API
onedrive    | Invalid response uri entered
```

To repeat: after the manual invocation on the command line, there is no `refresh.token` saved at that point in my config folder (i.e. before I then try and start the container) - I presume there should be, though?

**Comment by abraunegg** [comment]
@bossanova808 
> Yes, I saw that, but I think that is because no token is being saved in the initial step.

No .. that message *very* specific. It only ever gets presented when `--reauth` is used

> To repeat: after the manual invocation on the command line, there is no refresh.token saved at that point in my config folder (i.e. before I then try and start the container) - I presume there should be, though?

Correct - and, in usualy circumstances it should be there - but given your log output - it is really questionable as you what you are doing and what you are experiencing.

Check local file permissions, maybe even recreate your config Docker volume ... 

Enabling 2FA by your company does not cause a local file to not be written out ...

Please double check all your setup and process and commands.

**Comment by bossanova808** [comment]
Ok, so the problem _is_ the `re-auth` option it seems.

It that is run, _without_ a `refresh.token` in place, it creates the scenario I have described above (the new token is not saved).

I decided to treat things as a first run (so the docker command I used was thus simply:
```
    docker run -it --name onedrive \
    -v "/home/me/docker/config/onedrive":/onedrive/conf \
    -v "/media/store03/OneDrive":/onedrive/data \
    -e "ONEDRIVE_UID=1000" -e "ONEDRIVE_GID=1000" \
    driveone/onedrive:debian
 ```
That worked, a `refresh.token` was this time created, and actually started the container this time:

```
Initializing the Synchronization Engine ...
Initializing monitor ...
OneDrive monitor interval (seconds): 1800
```
I was then able to stop that, and `docker-compose up onedrive` successfully.

So something went wrong with the initial re-auth I guess, and the token was removed.  From then on using re-auth led to this issue, it seems.  All I am sure about is removing the re-auth - from the command line invocation only, no other changes - then allowed it to then work as expected.
